File: Ruby/048.rb
# summation(an) mod x = (a1 mod x + a2 mod x + a3 mod x + ... + an mod x) mod x
# i.e (a + b) mod x = (a % x + b % x) % x

MOD = 10**10

def modular_pow(base, exponent)
  result = 1
  while exponent > 0 do
    if exponent & 1 == 1
      result = (result * base) % MOD
    end
    exponent = exponent >> 1 # divide by 2
    base = (base * base) % MOD
  end
  result
end

puts (1..1000).inject{ |sum, n| (sum + modular_pow(n, n)) % MOD}

File: Cpp/048.cpp
#include<cstdio>
#define MOD 10000000000ll
using namespace std;

typedef unsigned long long ll;

ll modular_pow(ll base,ll exponent)
{
    ll result=1;
    while(exponent>0)
    {
        if((exponent&1)==1)
        {
            result=(result*base)%MOD;
        }
        exponent = exponent >> 1;   //divide by 2
        base=(base*base)%MOD;
    }
    return result;
}

int main()
{
    ll sum=0;
    ll i;
    for(i=1;i<=1000;i++)
    {
        sum=(sum+modular_pow(i,i))%MOD;
    }
    printf("%llu\n",sum);
}
